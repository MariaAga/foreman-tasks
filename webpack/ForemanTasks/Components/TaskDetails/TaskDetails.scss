.task-details-react {
  .progress-label-top-right {
    font-size: 11px;
    text-align: right;
  }
  a,
  button {
    margin-right: 3px;
  }
  .container {
    margin: 0;
  }
  .spin {
    -webkit-animation: spin 1s infinite linear;
    -moz-animation: spin 1s infinite linear;
    -o-animation: spin 1s infinite linear;
    animation: spin 1s infinite linear;
    -webkit-transform-origin: 50% 50%;
    transform-origin: 50% 50%;
    -ms-transform-origin: 50% 50%; /* IE 9 */
  }

  @-moz-keyframes spin {
    from {
      -moz-transform: rotate(0deg);
    }
    to {
      -moz-transform: rotate(360deg);
    }
  }

  @-webkit-keyframes spin {
    from {
      -webkit-transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(360deg);
    }
  }

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
  .dynflow-button > span {
    pointer-events: auto;
  }

  pre {
    white-space: pre;
    word-break: normal;
  }

  .param-name {
    display: inline-block;
    width: 10em;
  }
}
