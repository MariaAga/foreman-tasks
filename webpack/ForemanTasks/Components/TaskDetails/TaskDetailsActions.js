import API from 'foremanReact/API';
import { API_OPERATIONS } from 'foremanReact/redux/API';
import {
  showLoading,
  hideLoading,
} from 'foremanReact/components/Layout/LayoutActions';
import {
  FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST,
  FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  FOREMAN_TASK_DETAILS_FETCH_TASK_FAILURE,
  FOREMAN_TASK_DETAILS_STOP_PULLING,
  FOREMAN_TASK_DETAILS_START_PULLING,
  FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
  FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
} from './TaskDetailsConstants';

export const taskReloadStop = timeoutId => ({
  type: FOREMAN_TASK_DETAILS_STOP_PULLING,
  payload: { timeoutId },
});

export const taskReloadStart = (timeoutId, refetchTaskDetails, id) => ({
  type: FOREMAN_TASK_DETAILS_START_PULLING,
  payload: { timeoutId, refetchTaskDetails, id },
});

export const refetchTaskDetails = id => dispatch => {
  showLoading();
  dispatch(startRequest());
  reloadTasksDetails(id, dispatch);
};

const reloadTasksDetails = async (id, dispatch) => {
  try {
    const { data } = await API.get(`/foreman_tasks/api/tasks/${id}/details`);
    dispatch(requestSuccess(data));
  } catch (error) {
    dispatch(requestFailure(error));
    document.location.reload();
  } finally {
    hideLoading();
  }
};

export const fetchTaskDetails = id => dispatch => {
  showLoading();
  dispatch(startRequest());
  getTasksDetails(id, dispatch);
};

const getTasksDetails = async (id, dispatch) => {
  dispatch({
    type: API_OPERATIONS.GET,
    url: `/foreman_tasks/api/tasks/${id}/details`,
    outputType: 'FOREMAN_TASK_DETAILS_FETCH_TASK',
    /* onSuccess and OnFailure will be supported in the next API PR */
    onSuccess: hideLoading,
    onFailure: hideLoading,
  });
};

const startRequest = () => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST,
});

const requestSuccess = data => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  payload: data,
});

const requestFailure = error => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_FAILURE,
  payload: error,
});

export const toggleUnlockModal = () => ({
  type: FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
});
export const toggleForceUnlockModal = () => ({
  type: FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
});
