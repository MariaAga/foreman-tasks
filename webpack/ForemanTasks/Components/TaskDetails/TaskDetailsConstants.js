export const FOREMAN_TASK_DETAILS = 'FOREMAN_TASK_DETAILS';
export const FOREMAN_TASK_DETAILS_SUCCESS = 'FOREMAN_TASK_DETAILS_SUCCESS';

export const TASK_STEP_CANCEL = 'TASK_STEP_CANCEL';
