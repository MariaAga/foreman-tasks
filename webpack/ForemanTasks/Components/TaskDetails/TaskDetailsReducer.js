import Immutable from 'seamless-immutable';
import {
  FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  FOREMAN_TASK_DETAILS_STOP_PULLING,
  FOREMAN_TASK_DETAILS_START_PULLING,
  FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
  FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
} from './TaskDetailsConstants';

const initialState = Immutable({});

export default (state = initialState, action) => {
  const { type, payload } = action;

  switch (type) {
    case FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS:
      return state.merge({
        ...payload,
      });
    case FOREMAN_TASK_DETAILS_STOP_PULLING:
      if (payload.timeoutId) {
        clearTimeout(payload.timeoutId);
      }
      return state.merge({ taskReload: false, timeoutId: null });
    case FOREMAN_TASK_DETAILS_START_PULLING:
      if (!payload.timeoutId) {
        payload.timeoutId = setInterval(
          () => payload.refetchTaskDetails(payload.id),
          5000
        );
      }
      return state.merge({ taskReload: true, timeoutId: payload.timeoutId });
    case FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL:
      return state.set('showUnlockModal', !state.showUnlockModal);
    case FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL:
      return state.set('showForceUnlockModal', !state.showForceUnlockModal);
    default:
      return state;
  }
};
