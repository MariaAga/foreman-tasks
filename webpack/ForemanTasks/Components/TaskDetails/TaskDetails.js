import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Tab, Tabs } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';
import { STATUS } from 'foremanReact/constants';
import MessageBox from 'foremanReact/components/common/MessageBox';
import Task from './Components/Task';
import RunningSteps from './Components/RunningSteps';
import Errors from './Components/Errors';
import Locks from './Components/Locks';
import Raw from './Components/Raw';
import { getTaskID } from './TasksDetailsHelper';
import { TaskSkeleton } from './Components/TaskSkeleton';

import './TaskDetails.scss';

class TaskDetails extends Component {
  componentDidMount() {
    const { timeoutId, refetchTaskDetails, fetchTaskDetails } = this.props;
    fetchTaskDetails(getTaskID(), timeoutId, refetchTaskDetails);
  }
  componentWillUnmount() {
    this.props.taskReloadStop(this.props.timeoutId);
  }
  taskProgressToggle = () => {
    const {
      timeoutId,
      refetchTaskDetails,
      loading,
      taskReloadStop,
      taskReloadStart,
    } = this.props;
    const id = getTaskID();
    if (timeoutId) {
      taskReloadStop(timeoutId);
    } else {
      taskReloadStart(timeoutId, refetchTaskDetails, id, loading);
    }
  };
  render() {
    const {
      externalId,
      startedAt,
      endedAt,
      label,
      input,
      output,
      executionPlan,
      failedSteps,
      runningSteps,
      locks,
      cancelStep,
      status,
      isData,
      APIerror,
    } = this.props;
    const id = getTaskID();
    const resumable = executionPlan ? executionPlan.state === 'paused' : false;
    const cancellable = executionPlan ? executionPlan.cancellable : false;
    const loading = status === STATUS.PENDING && !isData;

    // const loading = true;
    if (status === STATUS.ERROR) {
      return (
        <MessageBox
          key="tasks-table-error"
          icontype="error-circle-o"
          msg={__(`Could not receive data: ${APIerror && APIerror.message}`)}
        />
      );
    }
    return (
      <div className="task-details-react well">
        <Tabs defaultActiveKey={1} animation={false} id="task-details-tabs">
          <Tab eventKey={1} title={__('Task')}>
            {loading ? (
              <TaskSkeleton />
            ) : (
              <Task
                {...{
                  ...this.props,
                  cancellable,
                  resumable,
                  id,
                  status,
                  taskProgressToggle: this.taskProgressToggle,
                }}
              />
            )}
          </Tab>
          <Tab eventKey={2} disabled={loading} title={__('Running Steps')}>
            <RunningSteps
              runningSteps={runningSteps}
              id={id}
              cancelStep={cancelStep}
              taskReload={this.props.taskReload}
              taskProgressToggle={this.taskProgressToggle}
            />
          </Tab>
          <Tab eventKey={3} disabled={loading} title={__('Errors')}>
            <Errors executionPlan={executionPlan} failedSteps={failedSteps} />
          </Tab>
          <Tab eventKey={4} disabled={loading} title={__('Locks')}>
            <Locks locks={locks} />
          </Tab>
          <Tab eventKey={5} disabled={loading} title={__('Raw')}>
            <Raw
              {...{ id, label, startedAt, endedAt, input, output, externalId }}
            />
          </Tab>
        </Tabs>
      </div>
    );
  }
}

TaskDetails.propTypes = {
  label: PropTypes.string,
  fetchTaskDetails: PropTypes.func.isRequired,
  runningSteps: PropTypes.array,
  cancelStep: PropTypes.func.isRequired,
  taskReload: PropTypes.bool.isRequired,
  status: PropTypes.oneOf(Object.keys(STATUS)),
  APIerror: PropTypes.object,
  ...Task.propTypes,
  ...Errors.propTypes,
  ...Locks.propTypes,
  ...Raw.propTypes,
};
TaskDetails.defaultProps = {
  label: '',
  runningSteps: [],
  APIerror: null,
  status: STATUS.PENDING,
  ...Task.defaultProps,
  ...RunningSteps.defaultProps,
  ...Errors.defaultProps,
  ...Locks.defaultProps,
  ...Raw.defaultProps,
};

export default TaskDetails;
