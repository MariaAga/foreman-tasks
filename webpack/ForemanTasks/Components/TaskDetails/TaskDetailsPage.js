import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import PageLayout from 'foremanReact/pages/common/PageLayout/PageLayout';
import TaskDetails from './TaskDetails';
import { getTaskID } from './TasksDetailsHelper';

export const indexBreadcrumbItem = {
  caption: __('Tasks'),
  url: '/foreman_tasks/tasks',
};

export const breadcrumbTaskItem = task => ({
  caption: task.action || __('N/A'),
  url: `/foreman_tasks/tasks/${getTaskID()}/details`,
});

class TaskDetailsPage extends Component {
  componentDidMount() {
    const { timeoutId, refetchTaskDetails, fetchTaskDetails } = this.props;

    fetchTaskDetails(getTaskID(), timeoutId, refetchTaskDetails);
  }
  componentWillUnmount() {
    this.props.taskReloadStop(this.props.timeoutId);
  }
  render() {
    return (
      <PageLayout
        searchable={false}
        breadcrumbOptions={{
          isSwitchable: true,
          breadcrumbItems: [
            indexBreadcrumbItem,
            breadcrumbTaskItem(this.props),
          ],
          resource: {
            nameField: 'action',
            resourceUrl: '/foreman_tasks/api/tasks',
            switcherItemUrl: '/foreman_tasks/tasks/:id',
          },
        }}
      >
        <TaskDetails {...this.props} />
      </PageLayout>
    );
  }
}

TaskDetailsPage.propTypes = {
  fetchTaskDetails: PropTypes.func,
  refetchTaskDetails: PropTypes.func,
  taskReloadStop: PropTypes.func,
  timeoutId: PropTypes.number,
};

TaskDetailsPage.defaultProps = {
  fetchTaskDetails: () => null,
  refetchTaskDetails: () => null,
  taskReloadStop: () => null,
  timeoutId: null,
};
export default TaskDetailsPage;
