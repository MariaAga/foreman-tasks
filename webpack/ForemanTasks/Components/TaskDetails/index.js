import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import TaskDetails from './TaskDetails';
import * as taskDetailsActions from './TaskDetailsActions';
import * as taskActions from '../TaskActions';
import reducer from './TaskDetailsReducer';
import {
  selectEndedAt,
  selectStartAt,
  selectStartBefore,
  selectStartedAt,
  selectLocks,
  selectInput,
  selectOutput,
  selectResumable,
  selectCancellable,
  selectHelp,
  selectErrors,
  selectProgress,
  selectUsername,
  selectLabel,
  selectExecutionPlan,
  selectFailedSteps,
  selectRunningSteps,
  selectHasSubTasks,
  selectUsernamePath,
  selectAction,
  selectState,
  selectResult,
  selectTimeoutId,
  selectTaskReload,
  selectParentTask,
  selectExternalId,
  selectDynflowEnableConsole,
  selectCanEdit,
  selectStatus,
  selectError,
  selectIsData,
} from './TaskDetailsSelectors';

const mapStateToProps = state => ({
  startAt: selectStartAt(state),
  startBefore: selectStartBefore(state),
  startedAt: selectStartedAt(state),
  endedAt: selectEndedAt(state),
  input: selectInput(state),
  output: selectOutput(state),
  resumable: selectResumable(state),
  cancellable: selectCancellable(state),
  errors: selectErrors(state),
  progress: selectProgress(state),
  username: selectUsername(state),
  label: selectLabel(state),
  executionPlan: selectExecutionPlan(state),
  failedSteps: selectFailedSteps(state),
  runningSteps: selectRunningSteps(state),
  help: selectHelp(state),
  hasSubTasks: selectHasSubTasks(state),
  locks: selectLocks(state),
  usernamePath: selectUsernamePath(state),
  action: selectAction(state),
  state: selectState(state),
  result: selectResult(state),
  timeoutId: selectTimeoutId(state),
  taskReload: selectTaskReload(state),
  parentTask: selectParentTask(state),
  externalId: selectExternalId(state),
  dynflowEnableConsole: selectDynflowEnableConsole(state),
  canEdit: selectCanEdit(state),
  status: selectStatus(state),
  error: selectError(state),
  isData: selectIsData(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ ...taskActions, ...taskDetailsActions }, dispatch);

export const reducers = { taskDetails: reducer };

export default connect(mapStateToProps, mapDispatchToProps)(TaskDetails);
