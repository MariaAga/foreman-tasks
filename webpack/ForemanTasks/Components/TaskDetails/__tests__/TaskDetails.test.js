import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import TaskDetails from '../TaskDetails';
import { minProps } from './TaskDetails.fixtures';

const fixtures = {
  'render with loading Props': { ...minProps, status: 'PENDING' },
  'render with min Props': minProps,
};

delete window.location;
window.location = new URL(
  'https://foreman.com/foreman_tasks/tasks/a15dd820-32f1-4ced-9ab7-c0fab8234c47/'
);

describe('TaskDetails', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TaskDetails, fixtures));
});
