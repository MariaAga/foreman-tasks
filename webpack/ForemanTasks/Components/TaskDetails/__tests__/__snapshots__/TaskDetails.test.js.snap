// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDetails rendering render with error Props 1`] = `
<MessageBox
  icontype="error-circle-o"
  key="tasks-table-error"
  msg="Could not receive data: some-error"
/>
`;

exports[`TaskDetails rendering render with loading Props 1`] = `
<div
  className="task-details-react well"
>
  <ForwardRef
    animation={false}
    defaultActiveKey={1}
    id="task-details-tabs"
  >
    <Tab
      eventKey={1}
      title="Task"
    >
      <TaskSkeleton />
    </Tab>
    <Tab
      disabled={true}
      eventKey={2}
      title="Running Steps"
    >
      <RunningSteps
        cancelStep={[MockFunction]}
        id="a15dd820-32f1-4ced-9ab7-c0fab8234c47"
        runningSteps={Array []}
        taskProgressToggle={[Function]}
        taskReload={false}
      />
    </Tab>
    <Tab
      disabled={true}
      eventKey={3}
      title="Errors"
    >
      <Errors
        executionPlan={Object {}}
        failedSteps={Array []}
      />
    </Tab>
    <Tab
      disabled={true}
      eventKey={4}
      title="Locks"
    >
      <Locks
        locks={Array []}
      />
    </Tab>
    <Tab
      disabled={true}
      eventKey={5}
      title="Raw"
    >
      <Raw
        endedAt=""
        externalId=""
        id="a15dd820-32f1-4ced-9ab7-c0fab8234c47"
        input={Array []}
        label=""
        output={Object {}}
        startedAt=""
      />
    </Tab>
  </ForwardRef>
</div>
`;

exports[`TaskDetails rendering render with min Props 1`] = `
<div
  className="task-details-react well"
>
  <ForwardRef
    animation={false}
    defaultActiveKey={1}
    id="task-details-tabs"
  >
    <Tab
      eventKey={1}
      title="Task"
    >
      <Task
        APIerror={null}
        action=""
        canEdit={false}
        cancelStep={[MockFunction]}
        cancelTaskRequest={[Function]}
        cancellable={false}
        dynflowEnableConsole={false}
        endedAt=""
        errors={Array []}
        executionPlan={Object {}}
        externalId=""
        failedSteps={Array []}
        fetchTaskDetails={
          [MockFunction] {
            "calls": Array [
              Array [
                "a15dd820-32f1-4ced-9ab7-c0fab8234c47",
                undefined,
                [MockFunction],
              ],
              Array [
                "a15dd820-32f1-4ced-9ab7-c0fab8234c47",
                undefined,
                [MockFunction],
              ],
              Array [
                "a15dd820-32f1-4ced-9ab7-c0fab8234c47",
                undefined,
                [MockFunction],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        hasSubTasks={false}
        help=""
        id="a15dd820-32f1-4ced-9ab7-c0fab8234c47"
        input={Array []}
        label=""
        locks={Array []}
        output={Object {}}
        parentTask=""
        progress={0}
        refetchTaskDetails={[MockFunction]}
        result="error"
        resumable={false}
        resumeTaskRequest={[Function]}
        runningSteps={Array []}
        startAt=""
        startBefore=""
        startedAt=""
        state=""
        status="RESOLVED"
        taskProgressToggle={[Function]}
        taskReload={false}
        taskReloadStart={[MockFunction]}
        taskReloadStop={[MockFunction]}
        username=""
        usernamePath=""
      />
    </Tab>
    <Tab
      disabled={false}
      eventKey={2}
      title="Running Steps"
    >
      <RunningSteps
        cancelStep={[MockFunction]}
        id="a15dd820-32f1-4ced-9ab7-c0fab8234c47"
        runningSteps={Array []}
        taskProgressToggle={[Function]}
        taskReload={false}
      />
    </Tab>
    <Tab
      disabled={false}
      eventKey={3}
      title="Errors"
    >
      <Errors
        executionPlan={Object {}}
        failedSteps={Array []}
      />
    </Tab>
    <Tab
      disabled={false}
      eventKey={4}
      title="Locks"
    >
      <Locks
        locks={Array []}
      />
    </Tab>
    <Tab
      disabled={false}
      eventKey={5}
      title="Raw"
    >
      <Raw
        endedAt=""
        externalId=""
        id="a15dd820-32f1-4ced-9ab7-c0fab8234c47"
        input={Array []}
        label=""
        output={Object {}}
        startedAt=""
      />
    </Tab>
  </ForwardRef>
</div>
`;
